File: skills/inbox-labeler/labels.py
#!/usr/bin/env python3
"""Inbox Labeler — local CRUD for labels.

A label is {label, type, instruction}, plus {required_labels,
recommended_labels} when it is derived. Labels live in data/labels.json at the
repository root. Every command prints JSON to stdout and exits non-zero with
{"error": ...} when something is wrong.

`label` is the label's only identifier. It is what the user reads, what other
labels reference, and how get, update and delete address it — there is no
separate name and no technical id. Labels are ordinary readable phrases and may
contain spaces: "Delivery arriving soon". They are matched case-insensitively
and stored with the spelling they were given.

`type` says how a label decides whether it applies. A `detection` label
inspects an email directly. A `derived` label interprets the email together
with the detection labels that matched it, naming them in `required_labels` and
`recommended_labels`. Another type is added by adding an entry to LABEL_TYPES
below — the operations in this file do not change.

`IL/` is Inbox Labeler's Gmail namespace — infrastructure, not part of a
label's identity — so it is added only when talking to Gmail, where
"Delivery arriving soon" resolves to "IL/Delivery arriving soon", spaces and
all. See gmail_label().

Usage:
    labels.py list
    labels.py get LABEL
    labels.py create --label LABEL --instruction TEXT [--type TYPE]
                     [--required-label LABEL ...] [--recommended-label LABEL ...]
    labels.py update LABEL [--label NEW] [--instruction TEXT] [--type TYPE]
                     [--required-label LABEL ...] [--recommended-label LABEL ...]
    labels.py delete LABEL
    labels.py attention [LABEL ...]
    labels.py policy ATTENTION --age AGE
    labels.py color ATTENTION

Passing --label to update renames the label and rewrites every reference to it.
"""

import argparse
import json
import sys
from pathlib import Path

# The store lives in the repository's data/ directory, not beside this script:
# the skill directory is reached through symlinks from .claude/skills/ and
# .agents/skills/, so resolve() first follows those back to the real file here
# in skills/inbox-labeler/, and data/ is two levels up from it.
STORE = Path(__file__).resolve().parents[2] / "data" / "labels.json"

# The Gmail namespace Inbox Labeler owns. Labels are stored without it and
# resolved through it on the way to Gmail.
GMAIL_NAMESPACE = "IL/"

# Inbox Labeler's own labels. They are internal state rather than anything a
# user models, so they are spelled lowercase — a convention that keeps them
# apart from the readable phrases users write — and reserved: no label may
# resolve to one of them. Matching is case-insensitive, so "Processed" is
# reserved too.
RESERVED_LABELS = {
    "processed": "Inbox Labeler's processing state",
    "no-match": "Inbox Labeler's evaluation outcome",
}

# What a label asks of the user. Attention is not a label and is not stored per
# message: the `attention` command computes it from the labels a message has and
# turns it into Gmail actions.
#
# The order is the priority aggregation follows, lowest first. `normal` is the
# absence of a request, so any label that does ask for something outranks it,
# and `high` outranks `none`.
ATTENTION_LEVELS = ("normal", "none", "high")
DEFAULT_ATTENTION = "normal"

# What the `attention` command does at each level. Fixed, not configurable.
ATTENTION_POLICIES = {
    "none": {"mark_read_after": "24h"},
    "normal": {},
    "high": {"star": True},
}

# Gmail label colors by attention level.
# Attention is the source of truth.
ATTENTION_COLORS = {
    "none": {"backgroundColor": "#cccccc", "textColor": "#000000"},
    "normal": {"backgroundColor": "#fce8b3", "textColor": "#000000"},
    "high": {"backgroundColor": "#efa093", "textColor": "#000000"},
}

DURATION_UNITS = {"s": 1, "m": 60, "h": 3600, "d": 86400, "w": 604800}

# Fields every label carries, whatever its type.
COMMON_FIELDS = ("label", "type", "attention")

# The label types Inbox Labeler understands. `fields` are the type's own
# required text fields, on top of COMMON_FIELDS; `references` are its lists of
# labels pointing at other labels. Adding a type means adding an entry here —
# the CRUD operations stay as they are.
LABEL_TYPES = {
    "detection": {
        "summary": "inspects an email directly and decides whether its Gmail label applies",
        "fields": ("instruction",),
        "references": (),
    },
    "derived": {
        "summary": "interprets an email together with the detection labels that matched it",
        "fields": ("instruction",),
        "references": ("required_labels", "recommended_labels"),
    },
}
DEFAULT_TYPE = "detection"

# Every reference field any type declares, for validating the ones that do not.
REFERENCE_FIELDS = tuple(
    field for spec in LABEL_TYPES.values() for field in spec.get("references", ())
)


class ValidationError(Exception):
    pass


# --- labels ----------------------------------------------------------------


def gmail_label(label):
    """Resolve a label to the Gmail label Inbox Labeler applies, spaces and all."""
    return GMAIL_NAMESPACE + label


def strip_namespace(label):
    """Remove exactly one leading `IL/` from a label, if it has one."""
    if label[: len(GMAIL_NAMESPACE)].lower() == GMAIL_NAMESPACE.lower():
        return label[len(GMAIL_NAMESPACE):]
    return label


def normalise(label):
    """Trim the ends and collapse inner runs of whitespace to single spaces."""
    return " ".join(str(label).split())


def same_label(one, other):
    """Labels identify case-insensitively, so "large amount" is "Large amount"."""
    return normalise(one).lower() == normalise(other).lower()


# --- attention -------------------------------------------------------------


def parse_duration(value):
    """Turn "24h" or "2d" into seconds."""
    text = str(value).strip().lower()
    unit = text[-1:]
    if unit not in DURATION_UNITS or not text[:-1].isdigit():
        raise ValidationError(
            "duration %r must be a whole number followed by %s, e.g. 24h"
            % (value, "/".join(sorted(DURATION_UNITS)))
        )
    return int(text[:-1]) * DURATION_UNITS[unit]


def effective_attention(triggered):
    """The highest-priority attention among the labels a message carries.

    One `high` makes it `high`; failing that, one `none` makes it `none`;
    otherwise it stays `normal`. Nothing else is consulted: not the email, not
    the clock. A message with no labels comes out at the default, so it is left
    alone.
    """
    levels = []
    for entry in triggered or []:
        level = (entry.get("attention") or DEFAULT_ATTENTION).strip().lower()
        if level not in ATTENTION_LEVELS:
            raise ValidationError(
                "label %r has an unknown attention %r" % (entry.get("label"), level)
            )
        levels.append(level)
    if not levels:
        return DEFAULT_ATTENTION
    return max(levels, key=ATTENTION_LEVELS.index)


def attention_actions(attention, age_seconds):
    """What the `attention` command should do to a message, from its level alone.

    `age_seconds` is how long ago the message was received. Returns action names:

        star        add Gmail's star
        mark_read   clear UNREAD

    Labels are deliberately not an input, and no action ever touches an `IL/`
    label.
    """
    if attention not in ATTENTION_LEVELS:
        raise ValidationError(
            "unknown attention %r — levels: %s" % (attention, ", ".join(ATTENTION_LEVELS))
        )
    policy = ATTENTION_POLICIES[attention]
    age = max(0, int(age_seconds))

    if policy.get("star"):
        return ["star"]
    mark_read_after = policy.get("mark_read_after")
    if mark_read_after and age >= parse_duration(mark_read_after):
        return ["mark_read"]
    return []


def attention_color(attention):
    """The Gmail color for an attention level — the one place this is decided.

    Raises for a level outside ATTENTION_LEVELS and, separately, for a level
    that has no entry in ATTENTION_COLORS, so a gap in the mapping fails loudly
    instead of a label silently going uncolored.
    """
    if attention not in ATTENTION_LEVELS:
        raise ValidationError(
            "unknown attention %r — levels: %s" % (attention, ", ".join(ATTENTION_LEVELS))
        )
    color = ATTENTION_COLORS.get(attention)
    if not color:
        raise ValidationError("no Gmail color is configured for attention %r" % attention)
    return color


# --- storage ---------------------------------------------------------------


def ordered(entry):
    """Return the label's fields in canonical order: label, type, instruction, …"""
    spec = LABEL_TYPES.get(entry.get("type"), {})
    order = list(COMMON_FIELDS) + list(spec.get("fields", ())) + list(spec.get("references", ()))
    result = {field: entry[field] for field in order if field in entry}
    result.update({field: value for field, value in entry.items() if field not in result})
    return result


def load_labels():
    """Read all labels, creating an empty store if needed.

    Stores written by an earlier version are normalised here: a technical `id`
    and a separate `name` are dropped, a leftover `IL/` prefix is stripped,
    whitespace is collapsed, a missing `type` becomes detection, and a derived
    label's reference lists are filled in. Reads never write — the next save
    persists the normalised values.
    """
    if not STORE.exists():
        save_labels([])
        return []
    try:
        data = json.loads(STORE.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise ValidationError("%s is not valid JSON: %s" % (STORE.name, exc))
    if not isinstance(data, list):
        raise ValidationError("%s must contain a JSON array" % STORE.name)
    for index, entry in enumerate(data):
        if not isinstance(entry, dict):
            continue
        entry.pop("id", None)
        if not entry.get("label") and entry.get("name"):
            entry["label"] = entry["name"]
        entry.pop("name", None)
        if isinstance(entry.get("label"), str):
            entry["label"] = normalise(strip_namespace(entry["label"].strip()))
        if not entry.get("type"):
            entry["type"] = DEFAULT_TYPE
        if not entry.get("attention"):
            entry["attention"] = DEFAULT_ATTENTION
        for field in LABEL_TYPES.get(entry["type"], {}).get("references", ()):
            entry[field] = [
                normalise(ref) for ref in entry.get(field) or [] if normalise(ref)
            ]
        data[index] = ordered(entry)
    return data


def save_labels(labels):
    STORE.parent.mkdir(parents=True, exist_ok=True)
    STORE.write_text(
        json.dumps(labels, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )


def find(labels, label):
    """Resolve a label by its text, case-insensitively."""
    for entry in labels:
        if same_label(entry.get("label", ""), label):
            return entry
    raise ValidationError("no label %r (use list to see them)" % normalise(label))


# --- validation ------------------------------------------------------------


def resolve_references(values, existing, own, field):
    """Check that every referenced label exists, and is a detection label.

    Returns the references spelled the way the labels they point at are spelled.
    """
    resolved = []
    for value in values or []:
        wanted = normalise(value)
        if not wanted:
            continue
        target = None
        for other in existing:
            if other is not own and same_label(other.get("label", ""), wanted):
                target = other
                break
        if target is None:
            known = ", ".join(
                '"%s"' % o.get("label")
                for o in existing
                if o is not own and o.get("type") == DEFAULT_TYPE
            )
            raise ValidationError(
                "%s references %r, which is not an existing label — detection labels: %s"
                % (field, wanted, known or "none yet")
            )
        if target.get("type") != DEFAULT_TYPE:
            raise ValidationError(
                '%s may only reference detection labels, and "%s" is a %s label'
                % (field, target.get("label"), target.get("type"))
            )
        if not any(same_label(target["label"], ref) for ref in resolved):
            resolved.append(target["label"])
    return resolved


def validate(entry, existing, own=None):
    """Validate a complete label. Returns a cleaned copy in canonical order."""
    cleaned = {field: normalise(entry.get(field) or "") for field in COMMON_FIELDS}
    cleaned["type"] = cleaned["type"].lower()
    cleaned["attention"] = (cleaned["attention"] or DEFAULT_ATTENTION).lower()
    label_type = cleaned["type"]
    label = cleaned["label"]

    if cleaned["attention"] not in ATTENTION_LEVELS:
        raise ValidationError(
            "unknown attention %r — levels, lowest priority first: %s"
            % (cleaned["attention"], ", ".join(ATTENTION_LEVELS))
        )

    if not label_type:
        raise ValidationError(
            "type must not be empty — supported types: %s" % ", ".join(sorted(LABEL_TYPES))
        )
    if label_type not in LABEL_TYPES:
        raise ValidationError(
            "unknown label type %r — supported types: %s"
            % (label_type, ", ".join(sorted(LABEL_TYPES)))
        )
    if own and own.get("type") and own["type"] != label_type:
        raise ValidationError(
            'a label\'s type is immutable: "%s" is a %s label and cannot become %s — '
            "create a new label instead" % (own.get("label"), own["type"], label_type)
        )

    if not label:
        raise ValidationError("label must not be empty")
    if label[: len(GMAIL_NAMESPACE)].lower() == GMAIL_NAMESPACE.lower():
        without = strip_namespace(label)
        raise ValidationError(
            "label is stored without the %s prefix, which Inbox Labeler adds for Gmail%s"
            % (GMAIL_NAMESPACE, (" — use %r instead of %r" % (without, label)) if without else "")
        )
    if label.startswith("/") or label.endswith("/") or "//" in label:
        raise ValidationError(
            "label must not start or end with '/' or contain '//' (it becomes %r in Gmail)"
            % gmail_label(label)
        )
    for reserved, purpose in RESERVED_LABELS.items():
        if same_label(label, reserved):
            raise ValidationError(
                '"%s" is a reserved system label: %s is %s, not something a label may model '
                "— choose a different label" % (label, gmail_label(reserved), purpose)
            )
    for other in existing:
        if other is not own and same_label(other.get("label", ""), label):
            raise ValidationError(
                'a label called "%s" already exists — labels are unique, ignoring case'
                % other["label"]
            )

    for field in LABEL_TYPES[label_type]["fields"]:
        value = (entry.get(field) or "").strip()
        if not value:
            raise ValidationError("%s must not be empty for a %s label" % (field, label_type))
        cleaned[field] = value

    references = LABEL_TYPES[label_type].get("references", ())
    for field in REFERENCE_FIELDS:
        if field in references:
            cleaned[field] = resolve_references(entry.get(field), existing, own, field)
        elif entry.get(field):
            supported = ", ".join(
                sorted(t for t, s in LABEL_TYPES.items() if field in s.get("references", ()))
            )
            raise ValidationError(
                "%s applies to %s labels, not to a %s label" % (field, supported, label_type)
            )

    return ordered(cleaned)


# --- operations ------------------------------------------------------------


def list_labels():
    return load_labels()


def get_label(label):
    return find(load_labels(), label)


def create_label(
    label,
    instruction,
    label_type=DEFAULT_TYPE,
    attention=DEFAULT_ATTENTION,
    required_labels=None,
    recommended_labels=None,
):
    labels = load_labels()
    entry = validate(
        {
            "label": label,
            "type": label_type,
            "attention": attention,
            "instruction": instruction,
            "required_labels": required_labels,
            "recommended_labels": recommended_labels,
        },
        labels,
    )
    labels.append(entry)
    save_labels(labels)
    return entry


def sync_match_store(change):
    """Apply a label change to the match store too, which is keyed by label text.

    `change` is called with the matches module. That module is imported here
    rather than at the top of the file because it imports this one; and its
    failures are re-raised as our own error because running this file as a script
    leaves matches.py holding a second copy of this module, whose ValidationError
    is a different class than the one main() catches.
    """
    import matches

    try:
        change(matches)
    except matches.ValidationError as exc:
        raise ValidationError(str(exc)) from None


def rename_references(labels, old, new):
    """Point every reference to `old` at `new` instead."""
    for other in labels:
        for field in LABEL_TYPES.get(other.get("type"), {}).get("references", ()):
            other[field] = [
                new if same_label(ref, old) else ref for ref in other.get(field, [])
            ]


def update_label(
    label,
    new_label=None,
    instruction=None,
    label_type=None,
    attention=None,
    required_labels=None,
    recommended_labels=None,
):
    changes = {
        "label": new_label,
        "instruction": instruction,
        "type": label_type,
        "attention": attention,
        "required_labels": required_labels,
        "recommended_labels": recommended_labels,
    }
    if all(value is None for value in changes.values()):
        raise ValidationError(
            "nothing to update: pass --label, --type, --attention, --instruction, "
            "--required-label or --recommended-label"
        )
    labels = load_labels()
    entry = find(labels, label)
    previous = entry["label"]
    merged = dict(entry)
    merged.update({field: value for field, value in changes.items() if value is not None})
    cleaned = validate(merged, labels, own=entry)
    entry.clear()
    entry.update(cleaned)
    if cleaned["label"] != previous:
        rename_references(labels, previous, cleaned["label"])
    save_labels(labels)
    if cleaned["label"] != previous:
        sync_match_store(lambda m: m.rename_label(previous, cleaned["label"]))
    return entry


def referencing_labels(labels, entry):
    """Labels whose reference lists point at this label."""
    found = []
    for other in labels:
        if other is entry:
            continue
        fields = LABEL_TYPES.get(other.get("type"), {}).get("references", ())
        if any(
            same_label(ref, entry.get("label", ""))
            for field in fields
            for ref in other.get(field, [])
        ):
            found.append(other)
    return found


def delete_label(label):
    for reserved, purpose in RESERVED_LABELS.items():
        if same_label(label, reserved):
            raise ValidationError(
                '"%s" is a reserved system label: %s is %s and is not stored as a label, '
                "so there is nothing to delete" % (normalise(label), gmail_label(reserved), purpose)
            )
    labels = load_labels()
    entry = find(labels, label)
    blocking = referencing_labels(labels, entry)
    if blocking:
        kinds = "/".join(sorted({o.get("type", "") for o in blocking}))
        raise ValidationError(
            'cannot delete %s label "%s": it is referenced by %s label%s %s — '
            "remove the reference first"
            % (
                entry.get("type"),
                entry.get("label"),
                kinds,
                "" if len(blocking) == 1 else "s",
                ", ".join('"%s"' % o.get("label") for o in blocking),
            )
        )
    labels.remove(entry)
    save_labels(labels)
    sync_match_store(lambda m: m.delete_label(entry["label"]))
    return entry


# --- cli -------------------------------------------------------------------


def main(argv=None):
    parser = argparse.ArgumentParser(description="Manage Inbox Labeler labels.")
    sub = parser.add_subparsers(dest="command", required=True)

    sub.add_parser("list", help="list all labels")

    show = sub.add_parser("get", help="show one label")
    show.add_argument("label", help="the label text, e.g. \"Delivery arriving soon\"")

    type_help = "label type (default: %s — %s)" % (
        DEFAULT_TYPE,
        ", ".join(sorted(LABEL_TYPES)),
    )
    attention_help = "what this label asks of the user (default: %s — %s)" % (
        DEFAULT_ATTENTION,
        ", ".join(ATTENTION_LEVELS),
    )
    label_help = (
        'the label, a readable phrase that may contain spaces, e.g. "Delivery arriving soon" '
        "(becomes IL/Delivery arriving soon in Gmail); no IL/ prefix"
    )
    required_help = (
        "derived labels only: a detection label that must have matched before this label is "
        "evaluated, given as its exact label text (repeatable)"
    )
    recommended_help = (
        "derived labels only: a detection label to offer as context when it matched, given as "
        "its exact label text (repeatable)"
    )

    create = sub.add_parser("create", help="create a label")
    create.add_argument("--label", required=True, help=label_help)
    create.add_argument(
        "--instruction", required=True, help="how Claude decides whether the label applies"
    )
    create.add_argument("--type", default=DEFAULT_TYPE, help=type_help)
    create.add_argument("--attention", default=DEFAULT_ATTENTION, help=attention_help)
    create.add_argument(
        "--required-label", action="append", dest="required_labels", metavar="LABEL",
        help=required_help,
    )
    create.add_argument(
        "--recommended-label", action="append", dest="recommended_labels", metavar="LABEL",
        help=recommended_help,
    )

    update = sub.add_parser("update", help="update a label")
    update.add_argument("label", help="the label to update, by its current text")
    update.add_argument(
        "--label", dest="new_label",
        help="rename the label to this text; every reference to it is updated too",
    )
    update.add_argument("--instruction")
    update.add_argument("--type", help=type_help)
    update.add_argument("--attention", help=attention_help)
    update.add_argument(
        "--required-label", action="append", dest="required_labels", metavar="LABEL",
        help=required_help + "; replaces the stored list",
    )
    update.add_argument(
        "--recommended-label", action="append", dest="recommended_labels", metavar="LABEL",
        help=recommended_help + "; replaces the stored list",
    )

    delete = sub.add_parser("delete", help="delete a label")
    delete.add_argument("label", help="the label text")

    level = sub.add_parser(
        "attention", help="the effective attention for the labels a message carries"
    )
    level.add_argument("labels", nargs="*", metavar="LABEL", help="the labels on the message")

    act = sub.add_parser(
        "policy", help="what the attention command should do to a message at this level"
    )
    act.add_argument("attention", choices=ATTENTION_LEVELS)
    act.add_argument("--age", required=True, help="how long ago it was received, e.g. 30h")

    col = sub.add_parser(
        "color", help="the Gmail color for an attention level"
    )
    col.add_argument("attention", choices=ATTENTION_LEVELS)

    args = parser.parse_args(argv)

    try:
        if args.command == "list":
            result = list_labels()
        elif args.command == "get":
            result = get_label(args.label)
        elif args.command == "create":
            result = create_label(
                args.label,
                args.instruction,
                args.type,
                args.attention,
                args.required_labels,
                args.recommended_labels,
            )
        elif args.command == "update":
            result = update_label(
                args.label,
                args.new_label,
                args.instruction,
                args.type,
                args.attention,
                args.required_labels,
                args.recommended_labels,
            )
        elif args.command == "attention":
            labels = load_labels()
            triggered, unknown = [], []
            for text in args.labels or []:
                match = next((e for e in labels if same_label(e["label"], text)), None)
                (triggered if match else unknown).append(match or normalise(text))
            result = {
                "attention": effective_attention(triggered),
                "from": {e["label"]: e["attention"] for e in triggered},
                "unknown": unknown,
            }
        elif args.command == "policy":
            result = {
                "attention": args.attention,
                "actions": attention_actions(args.attention, parse_duration(args.age)),
            }
        elif args.command == "color":
            result = {"attention": args.attention, "color": attention_color(args.attention)}
        else:
            result = delete_label(args.label)
    except ValidationError as exc:
        print(json.dumps({"error": str(exc)}, indent=2))
        return 1

    print(json.dumps(result, indent=2, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: web/lib/inbox/labels.ts
/**
 * What a label is, and every rule about one that does not need a database.
 *
 * This is the definition: what a label is, what spellings are accepted, and how
 * one is rejected. Every way into the product goes through it — the MCP tools, the
 * signed-in page, the store — so there is one answer to "is this a valid label"
 * rather than one per caller. `store/` adds only what needs the database.
 *
 * The field names are snake_case because they are the label's own field names,
 * the ones an MCP client passes and reads back, not this language's.
 * `lib/label-graph.ts` describes the same shape for the web UI; the two agree because
 * they are both descriptions of one document.
 */

/** The Gmail namespace InboxLabeler owns. Labels are stored without it. */
export const GMAIL_NAMESPACE = "IL/";

/**
 * What a label asks of the user, lowest priority first.
 *
 * The order is the priority aggregation follows: `normal` is the absence of a
 * request, so any label that does ask for something outranks it, and `high`
 * outranks `none`. Nothing here aggregates — `inbox-labeler-attention` does, per
 * message — but the order is part of the vocabulary and belongs with it.
 */
export const ATTENTION_LEVELS = ["normal", "none", "high"] as const;
export type Attention = (typeof ATTENTION_LEVELS)[number];
export const DEFAULT_ATTENTION: Attention = "normal";

/** How a label decides whether it applies. */
export const LABEL_TYPES = ["detection", "derived"] as const;
export type LabelType = (typeof LABEL_TYPES)[number];
export const DEFAULT_TYPE: LabelType = "detection";

/**
 * What a detection label's fact *is*, which is a different question from how it
 * was found.
 *
 * A category answers "what kind of email is this" — `Invoice`, `Delivery`,
 * `Newsletter`, `Travel`. An attribute answers "what does this email contain,
 * indicate or require" — `Action required`, `Deadline`, `Large amount`. The
 * difference is which question the label's own instruction is deciding, not which
 * words its name is made of.
 *
 * Neither is exclusive and neither is a bucket: one email can match several
 * categories and several attributes, and matching remains one independent decision
 * per label. Role changes how a matched fact is *read*, never whether it matched.
 *
 * Only detection labels have one. A derived label is already an interpretation of
 * detection facts, so asking what kind of fact it is has no answer.
 */
export const DETECTION_ROLES = ["category", "attribute"] as const;
export type DetectionRole = (typeof DETECTION_ROLES)[number];

/**
 * InboxLabeler's own two labels.
 *
 * Internal state rather than anything a user models, spelled lowercase — the
 * convention that keeps them apart from the readable phrases users write — and
 * reserved, so no label may resolve to one of them.
 */
export const RESERVED_LABELS: Readonly<Record<string, string>> = {
  processed: "InboxLabeler's processing state",
  "no-match": "InboxLabeler's evaluation outcome",
};

/** Which reference lists a type carries. A detection label carries none. */
export const REFERENCE_FIELDS = ["required_labels", "recommended_labels"] as const;
export type ReferenceField = (typeof REFERENCE_FIELDS)[number];

/** A label, exactly as the store holds one and `get_labels` returns it. */
export type Label = {
  label: string;
  type: LabelType;
  /**
   * Detection labels only: whether the fact is a category or an attribute.
   *
   * Optional, and deliberately so. Every detection label created from now on has
   * one — `checkRole` refuses a new one without it — but accounts predate the
   * distinction, and their labels are read and used exactly as before until
   * somebody decides with the user which role each plays. Absent means unmodelled,
   * never a default.
   */
  role?: DetectionRole;
  attention: Attention;
  instruction: string;
  /** Derived labels only: the detection labels that must all have matched. */
  required_labels?: string[];
  /** Derived labels only: detection labels offered as context when they matched. */
  recommended_labels?: string[];
};

/**
 * Something the caller got wrong, phrased for them.
 *
 * Its own class so a store can tell a rejected label from a database failure,
 * and so the MCP tools can answer one as the caller's problem and the other as
 * ours. The messages are the ones the product uses everywhere, so a client and a
 * reader of the skills meet one vocabulary of complaints rather than two.
 */
export class LabelError extends Error {
  override readonly name = "LabelError";
}

// --- identity --------------------------------------------------------------

/** Trim the ends and collapse inner runs of whitespace to single spaces. */
export function normalise(label: unknown): string {
  return String(label ?? "").split(/\s+/).filter(Boolean).join(" ");
}

/** Labels identify case-insensitively, so "large amount" is "Large amount". */
export function sameLabel(one: string, other: string): boolean {
  return normalise(one).toLowerCase() === normalise(other).toLowerCase();
}

/** Remove exactly one leading `IL/`, if present. */
export function stripNamespace(label: string): string {
  return label.slice(0, GMAIL_NAMESPACE.length).toLowerCase() === GMAIL_NAMESPACE.toLowerCase()
    ? label.slice(GMAIL_NAMESPACE.length)
    : label;
}

/** Resolve a label to the Gmail label InboxLabeler applies, spaces and all. */
export function gmailLabel(label: string): string {
  return GMAIL_NAMESPACE + label;
}

// --- the rules that need nothing but the value -----------------------------

/**
 * Checks a label's own text, returning it normalised.
 *
 * Everything here is decidable from the string alone. Uniqueness and references
 * need to know what else exists, so they belong to the store, which has the
 * transaction to check them in.
 */
export function checkLabelText(value: unknown): string {
  const label = normalise(value);

  if (!label) throw new LabelError("label must not be empty");

  if (label.slice(0, GMAIL_NAMESPACE.length).toLowerCase() === GMAIL_NAMESPACE.toLowerCase()) {
    const without = stripNamespace(label);
    throw new LabelError(
      `label is stored without the ${GMAIL_NAMESPACE} prefix, which InboxLabeler adds for Gmail` +
        (without ? ` — use "${without}" instead of "${label}"` : ""),
    );
  }
  if (label.startsWith("/") || label.endsWith("/") || label.includes("//")) {
    throw new LabelError(
      `label must not start or end with '/' or contain '//' (it becomes "${gmailLabel(label)}" in Gmail)`,
    );
  }
  for (const [reserved, purpose] of Object.entries(RESERVED_LABELS)) {
    if (sameLabel(label, reserved)) {
      throw new LabelError(
        `"${label}" is a reserved system label: ${gmailLabel(reserved)} is ${purpose}, ` +
          "not something a label may model — choose a different label",
      );
    }
  }
  return label;
}

/** Checks an attention level, defaulting an absent one. */
export function checkAttention(value: unknown): Attention {
  const attention = normalise(value).toLowerCase() || DEFAULT_ATTENTION;
  if (!(ATTENTION_LEVELS as readonly string[]).includes(attention)) {
    throw new LabelError(
      `unknown attention "${attention}" — levels, lowest priority first: ${ATTENTION_LEVELS.join(", ")}`,
    );
  }
  return attention as Attention;
}

/** Checks a label type, defaulting an absent one. */
export function checkType(value: unknown): LabelType {
  const type = normalise(value).toLowerCase() || DEFAULT_TYPE;
  if (!(LABEL_TYPES as readonly string[]).includes(type)) {
    throw new LabelError(
      `unknown label type "${type}" — supported types: ${[...LABEL_TYPES].sort().join(", ")}`,
    );
  }
  return type as LabelType;
}

/**
 * Checks a detection label's role.
 *
 * `existing` is the label being changed, or `undefined` when one is being created,
 * and that parameter is the whole of the legacy story:
 *
 *   - creating a detection label without a role is refused, so the distinction
 *     holds for everything new;
 *   - changing one that has none, without mentioning role, keeps none — editing an
 *     instruction must not turn into a modelling decision the caller was never
 *     asked to make;
 *   - passing a role, on create or update, stores it. Unlike `type`, it may be
 *     changed as often as the user changes their mind.
 *
 * A role on a derived label is refused rather than dropped, the same way a
 * reference list on a detection label is: it is a request the caller expects to
 * have an effect.
 */
export function checkRole(
  value: unknown,
  type: LabelType,
  existing: Label | undefined,
): DetectionRole | undefined {
  const given = normalise(value).toLowerCase();

  if (type !== "detection") {
    if (given) throw new LabelError(`role applies to detection labels, not to a ${type} label`);
    return undefined;
  }

  if (!given) {
    if (!existing) {
      throw new LabelError(
        "a detection label must say which role its fact plays: " +
          `${[...DETECTION_ROLES].sort().join(" or ")} — a category is what kind of email this ` +
          "is (Invoice, Delivery), an attribute is what the email contains, indicates or " +
          "requires (Action required, Deadline)",
      );
    }
    // Unmodelled stays unmodelled. Nothing here guesses.
    return existing.role;
  }

  if (!(DETECTION_ROLES as readonly string[]).includes(given)) {
    throw new LabelError(
      `unknown role "${given}" — roles: ${[...DETECTION_ROLES].sort().join(", ")}`,
    );
  }
  return given as DetectionRole;
}

/** Checks an instruction, which no type may leave empty. */
export function checkInstruction(value: unknown, type: LabelType): string {
  const instruction = String(value ?? "").trim();
  if (!instruction) {
    throw new LabelError(`instruction must not be empty for a ${type} label`);
  }
  return instruction;
}

/**
 * Rejects reference lists on a type that has none.
 *
 * A detection label with `required_labels` is not a detection label with a
 * harmless extra field — it is a request the caller expects to have an effect,
 * and silently dropping it would be the wrong kindness.
 */
export function checkReferencesAllowed(type: LabelType, given: Partial<Record<ReferenceField, unknown>>): void {
  if (type === "derived") return;
  for (const field of REFERENCE_FIELDS) {
    const value = given[field];
    if (Array.isArray(value) ? value.length > 0 : Boolean(value)) {
      throw new LabelError(`${field} applies to derived labels, not to a ${type} label`);
    }
  }
}

/** A label's type never changes; a new label is the way to change one. */
export function checkTypeUnchanged(existing: Label, wanted: LabelType): void {
  if (existing.type !== wanted) {
    throw new LabelError(
      `a label's type is immutable: "${existing.label}" is a ${existing.type} label and cannot ` +
        `become ${wanted} — create a new label instead`,
    );
  }
}

// --- shape -----------------------------------------------------------------

/**
 * A label with its fields in canonical order, and reference lists only where
 * they belong.
 *
 * The order is the one the product documents: label, type, role, attention,
 * instruction, then the references. It is what the MCP endpoint returns and what
 * the skills describe — a client and a reader should not see two orders for one
 * label.
 *
 * `role` sits beside `type` because it refines it, and is left out entirely rather
 * than sent as null when there is none: a derived label has no role to have, and a
 * detection label from before the distinction has not been given one yet. An absent
 * key says that; `null` would look like a decision.
 */
export function orderLabel(entry: Label): Label {
  const ordered: Label = {
    label: entry.label,
    type: entry.type,
    ...(entry.type === "detection" && entry.role ? { role: entry.role } : {}),
    attention: entry.attention,
    instruction: entry.instruction,
  };
  if (entry.type === "derived") {
    ordered.required_labels = entry.required_labels ?? [];
    ordered.recommended_labels = entry.recommended_labels ?? [];
  }
  return ordered;
}

/** Labels alphabetically, ignoring case: the order the store reads best in. */
export function byLabel(labels: readonly Label[]): Label[] {
  return [...labels].sort((a, b) =>
    a.label.toLowerCase() < b.label.toLowerCase() ? -1 : a.label.toLowerCase() > b.label.toLowerCase() ? 1 : 0,
  );
}

// --- messages the store needs ---------------------------------------------
//
// Built here rather than at the query, so that every complaint the product makes
// about a label is written in one place and worded the same way each time.

export function labelNotFound(label: string): LabelError {
  return new LabelError(`no label "${normalise(label)}" (use get_labels to see them)`);
}

export function labelExists(existing: string): LabelError {
  return new LabelError(
    `a label called "${existing}" already exists — labels are unique, ignoring case`,
  );
}

export function referenceMissing(field: ReferenceField, wanted: string, detection: readonly string[]): LabelError {
  return new LabelError(
    `${field} references "${wanted}", which is not an existing label — detection labels: ` +
      (detection.length ? detection.map((one) => `"${one}"`).join(", ") : "none yet"),
  );
}

export function referenceNotDetection(field: ReferenceField, target: Label): LabelError {
  return new LabelError(
    `${field} may only reference detection labels, and "${target.label}" is a ${target.type} label`,
  );
}

export function deleteBlocked(entry: Label, blocking: readonly Label[]): LabelError {
  const kinds = [...new Set(blocking.map((one) => one.type))].sort().join("/");
  return new LabelError(
    `cannot delete ${entry.type} label "${entry.label}": it is referenced by ${kinds} ` +
      `label${blocking.length === 1 ? "" : "s"} ` +
      `${blocking.map((one) => `"${one.label}"`).join(", ")} — remove the reference first`,
  );
}

export function deleteReserved(label: string): LabelError {
  const purpose = RESERVED_LABELS[normalise(label).toLowerCase()];
  return new LabelError(
    `"${normalise(label)}" is a reserved system label: ${gmailLabel(normalise(label).toLowerCase())} ` +
      `is ${purpose}, and is not stored as a label, so there is nothing to delete`,
  );
}

export function nothingToUpdate(): LabelError {
  return new LabelError(
    "nothing to update: pass a new label, type, role, attention, instruction, required_labels or recommended_labels",
  );
}
